core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -0,0 +1,188 @@
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  /**
+   * Creates an instance of remote {@link BlockInStream} that reads from a remote worker.
+   *
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createRemoteBlockInStream(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, BlockStoreContext context, InStreamOptions options)
+    throws IOException {
+    Closer closer = Closer.create();
+
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createWorkerClient(workerNetAddress));
+      LockBlockResult lockBlockResult = blockWorkerClient.lockBlock(blockId);
+      PacketReader.Factory factory = new NettyPacketReader.Factory(
+          blockWorkerClient.getDataServerAddress(), blockId, lockBlockResult.getLockId(),
+          blockWorkerClient.getSessionId());
+      return new BlockInStream(factory, blockId, blockSize, blockWorkerClient, options);
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      mBlockWorkerClient.unlockBlock(mBlockId);
+    } catch (Throwable e) { // must catch Throwable
+      throw mCloser.rethrow(e); // IOException will be thrown as-is
+    } finally {
+      mCloser.close();
+    }
+  }
+
+  @Override
+  public long remaining() {
+    return mInputStream.remaining();
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    mInputStream.seek(pos);
+  }
+
+  @Override
+  public int read(long pos, byte[] b, int off, int len) throws IOException {
+    return mInputStream.read(pos, b, off, len);
+  }
+
+  @Override
+  public InetSocketAddress location() {
+    return mBlockWorkerClient.getDataServerAddress();
+  }
+
+  @Override
+  public boolean isLocal() {
+    return mLocal;
+  }
+
+  /**
+   * Creates an instance of {@link BlockInStream}.
+   *
+   * @param packetReaderFactory the packet reader factory
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param blockWorkerClient the block worker client
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   */
+  private BlockInStream(PacketReader.Factory packetReaderFactory, long blockId, long blockSize,
+      BlockWorkerClient blockWorkerClient, InStreamOptions options) throws IOException {
+    super(new PacketInStream(packetReaderFactory, blockId, blockSize));
+
+    assert in instanceof PacketInStream;
```

Comment:
Done. Not sure what is the convention in Alluxio. I used "assert" because some code in Alluxio does this when checking the type of object. I thought the motivation is to save CPU cycles as "assert" is not enabled by default in production but Preconditions is.